src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Can't add value in position: " + index
+                    + ". Index is invalid.");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        checkElementIndex(index);
+        Node<T> currentNode;
+        if (index < (size << 1)) {
+            currentNode = first;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = last;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private void insertToBeginning(T value) {
+        Node<T> afterNode = first;
+        Node<T> newNode = new Node<>(null, value, afterNode);
+        if (afterNode != null) {
+            afterNode.prev = newNode;
+        } else {
+            last = newNode;
+        }
+        first = newNode;
+    }
+
+    private void insertBeforeIndex(T value, int index) {
+        Node<T> linkBefore = findNodeByIndex(index - 1);
+        Node<T> linkAfter = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(linkBefore, value, linkAfter);
+        linkBefore.next = newNode;
+        if (linkAfter != null) {
+            linkAfter.prev = newNode;
+        } else {
+            last = newNode;
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index:" + index);
+        }
+    }
+
+    private boolean equalsElement(T value, Object object) {
+        if (object == value) {
+            return true;
+        }
+        if (object == null || value == null) {
+            return false;
+        }
+        if (object.getClass() == value.getClass()) {
+            T current = (T) object;
+            return current == value || (current != null
+                    && current.equals(value));
+        }
         return false;
     }
+
+    private void unlink(Node<T> node) {
+        Node<T> prev = node.prev;
+        Node<T> next = node.next;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+        }
+    }
+
+    public static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
```

Comment:
The `Node<T>` class should be moved to the end of the `MyLinkedList` class. This change is necessary to meet the checklist requirement and improve code readability and maintainability.